File: Detect cycle in a directed graph - GFG/detect-cycle-in-a-directed-graph.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
class Solution {
  public:
    // Function to detect cycle in a directed graph.
    bool checkcycle(int node,vector<int>&vis,vector<int>&pathvis,vector<int>adj[])
    {
        vis[node]=1;
        pathvis[node]=1;
        for(auto it:adj[node])
        {
            if(!vis[it])
            {
                if(checkcycle(it,vis,pathvis,adj))return true;
            }else if(pathvis[it])
            {
                return true;
            }
        }
        pathvis[node]=0;
        return false;
    }
    bool isCyclic(int V, vector<int> adj[]) {
        // code here
        vector<int> vis(V,0);
        vector<int> pathvis(V,0);
        
        for(int i=0;i<V;i++)
        {
            if(!vis[i])
            if(checkcycle(i,vis,pathvis,adj))return true;
        }
        return false;
    }
};

//{ Driver Code Starts.

int main() {

    int t;
    cin >> t;
    while (t--) {
        int V, E;
        cin >> V >> E;

        vector<int> adj[V];

        for (int i = 0; i < E; i++) {
            int u, v;
            cin >> u >> v;
            adj[u].push_back(v);
        }

        Solution obj;
        cout << obj.isCyclic(V, adj) << "\n";
    }

    return 0;
}

// } Driver Code Ends

File: Detect cycle in a directed graph - GFG/detect-cycle-in-a-directed-graph.py
#User function Template for python3


class Solution:
    
    #Function to detect cycle in a directed graph.
    def isCyclic(self, V, adj):
        # code here
        
        # pathvis=[0]*V
        # vis=[0]*V
        # def dfs(node):
        #     vis[node]=1
        #     pathvis[node]=1
        #     for i in adj[node]:
        #         if not vis[i]:
        #             if(dfs(i)):
        #                 return True
        #         elif pathvis[i]:
        #             return True
        #     pathvis[node]=0
        #     return False
        # for i in range(V):
        #     if dfs(i)==True:
        #         return True
        # return False
        
        # TOPO
        
        ind=[0]*V
        q=[]
        
        for i in range(V):
            for j in adj[i]:
                ind[j]+=1
        for i in range(V):
            if ind[i]==0:
                q.append(i)
        stack=[] 
        while q:
            node=q.pop(0)
            stack.append(node)
            for i in adj[node]:
                ind[i]-=1
                if ind[i]==0:
                    q.append(i)
        if len(stack)==V:
            return False
        return True


#{ 
 # Driver Code Starts
#Initial Template for Python 3

import sys
sys.setrecursionlimit(10**6)
        
if __name__ == '__main__':
    t = int(input())
    for i in range(t):
        V,E = list(map(int, input().strip().split()))
        adj = [[] for i in range(V)]
        for i in range(E):
            a,b = map(int,input().strip().split())
            adj[a].append(b)
        ob = Solution()
        
        if ob.isCyclic(V, adj):
            print(1)
        else:
            print(0)
# } Driver Code Ends